modules/sc-mesh-secure-deployment/src/2_0/features/jamming/util.py

```diff
@@ -0,0 +1,238 @@
+
+    :param channel: The channel number.
+    :return: The frequency.
+    """
+    return CH_TO_FREQ[channel]
+
+
+def map_freq_to_channel(freq: int) -> int:
+    """
+    Maps the given frequency to its corresponding channel number int.
+
+    :param freq: The frequency value.
+    :return: The frequency.
+    """
+    return FREQ_TO_CH[freq]
+
+
+def load_sample_data(args: Options = Options()) -> Tuple[str, pd.DataFrame]:
+    """
+    Load sample data from a CSV file based on a randomly chosen sample type.
+
+    :param args: An Options object containing the command-line arguments.
+    :return: A tuple containing the message describing the sample data and a pandas DataFrame with the loaded data.
+    """
+    sample_type_to_dir = {
+        'floor': 'sample/floor/',
+        'jamming': 'sample/jamming/',
+        'background': 'sample/background/'
+    }
+
+    # Choose a random sample_type from the dictionary
+    sample_type = random.choice(list(sample_type_to_dir.keys()))
+    if sample_type not in sample_type_to_dir:
+        raise ValueError("Invalid sample_type provided. Choose from 'floor', 'background', or 'jamming'.")
+
+    csv_files = pathlib.Path(sample_type_to_dir[sample_type]).glob('*.csv')
+    path = random.choice(list(csv_files))
+    message = f'{sample_type.capitalize()} data {path}'
+
+    # load the data into a pandas DataFrame
+    data = pd.read_csv(path)
+    print(f"Sampled {sample_type} data") if args.debug else None
+    return message, data
+
+
+def get_frequency_quality(freq_quality: np.ndarray, probs: np.ndarray, frequencies: np.ndarray) -> Tuple[bool, dict]:
+    """
+    Get frequency quality information based on frequency quality and interference probabilities.
+
+    param freq_quality: NumPy array of frequency quality values.
+    param probs: NumPy array of interference probabilities.
+    param frequencies: NumPy array of frequencies to assess.
+
+    return: A Tuple containing a boolean indicating if jamming was detected and a dictionary mapping frequencies to quality values.
+    """
+    jamming_detected = False
+    freq_quality_dict = {}
+
+    # Check if passed params freq_quality, probs, frequencies are empty
+    if freq_quality.size == 0 or frequencies.size == 0 or probs.size == 0:
+        return False, {}
+
+    for i, freq in enumerate(frequencies):
+        index = np.argmax(probs[i])
+        freq_quality_dict[str(freq)] = float(freq_quality[i])
+        if index == 0:
+            pred = 'floor like interference'
+        elif index == 1:
+            pred = 'wireless lab. like interference'
+        elif index == 2:
+            pred = 'crypto/finance like interference'
+        else:
+            pred = 'jamming'
+            jamming_detected = True
+
+        print(freq, freq_quality[i], np.argmax(probs[i]), pred) if args.debug else None
+
+    return jamming_detected, freq_quality_dict
+
+
+# Frequency switch related functions
+def run_command(command, error_message) -> None:
```

Comment:
Not good practice as PID is lost.

Reply:
For processes that don't require pid tracking, for example displaying a list of available wireless devices with `iw dev`, there is typically no need for PID control. The majority of shell commands in the code are one-time executions that do not require continuous monitoring as they execute in the moment they are invoked and conclude their tasks without ongoing background processes. Based on your comments, PID control would mainly be necessary for processes such as the starting/killing wpa supplicant and osf. Is this in line with what you are suggesting?

Reply:
Yes, support for processes which you are later killing or stopping.   Support also for cases when some other service is already responsible for radio control.

Reply:
How this is taken account with the latest commit?  Let's say that if we have a separate service starting mesh in a device boot and that service shares information to other systems (topology, network perf etc..).   Then this jamming avoidance is started --> what happens?  Radio is controller from two different services?